src/main/resources/jupyter/init-actions.sh

```diff
@@ -160,130 +164,142 @@ if [[ "${ROLE}" == 'Master' ]]; then
+          log 'Installing Jupyter combined extension [$ext]...'
+          log $ext
+          if [[ $ext == 'gs://'* ]]; then
+            gsutil cp $ext /etc
+            JUPYTER_EXTENSION_ARCHIVE=`basename $ext`
+            docker cp /etc/${JUPYTER_EXTENSION_ARCHIVE} ${JUPYTER_SERVER_NAME}:${JUPYTER_HOME}/${JUPYTER_EXTENSION_ARCHIVE}
+            retry 3 docker exec -u root -e PIP_USER=false ${JUPYTER_SERVER_NAME} ${JUPYTER_SCRIPTS}/extension/jupyter_install_combined_extension.sh ${JUPYTER_EXTENSION_ARCHIVE}
+          else
+            retry 3 docker exec -u root -e PIP_USER=false ${JUPYTER_SERVER_NAME} ${JUPYTER_SCRIPTS}/extension/jupyter_pip_install_combined_extension.sh $ext
+          fi
+        done
+      fi
+
+      # If a custom.js was specified, copy it into the jupyter docker container.
+      if [ ! -z ${JUPYTER_CUSTOM_JS_URI} ] ; then
+        log 'Installing Jupyter custom.js...'
+        gsutil cp ${JUPYTER_CUSTOM_JS_URI} /etc
+        JUPYTER_CUSTOM_JS=`basename ${JUPYTER_CUSTOM_JS_URI}`
+        retry 3 docker exec ${JUPYTER_SERVER_NAME} mkdir -p ${JUPYTER_USER_HOME}/.jupyter/custom
+        docker cp /etc/${JUPYTER_CUSTOM_JS} ${JUPYTER_SERVER_NAME}:${JUPYTER_USER_HOME}/.jupyter/custom/
+      fi
+
+      # If a google_sign_in.js was specified, copy it into the jupyter docker container.
+      if [ ! -z ${JUPYTER_GOOGLE_SIGN_IN_JS_URI} ] ; then
+        log 'Installing Google sign in extension...'
+        gsutil cp ${JUPYTER_GOOGLE_SIGN_IN_JS_URI} /etc
+        JUPYTER_GOOGLE_SIGN_IN_JS=`basename ${JUPYTER_GOOGLE_SIGN_IN_JS_URI}`
+        retry 3 docker exec ${JUPYTER_SERVER_NAME} mkdir -p ${JUPYTER_USER_HOME}/.jupyter/custom
+        docker cp /etc/${JUPYTER_GOOGLE_SIGN_IN_JS} ${JUPYTER_SERVER_NAME}:${JUPYTER_USER_HOME}/.jupyter/custom/
+      fi
+
+      if [ ! -z ${JUPYTER_NOTEBOOK_CONFIG_URI} ] ; then
+        log 'Copy Jupyter notebook config...'
+        gsutil cp ${JUPYTER_NOTEBOOK_CONFIG_URI} /etc
+        JUPYTER_NOTEBOOK_CONFIG=`basename ${JUPYTER_NOTEBOOK_CONFIG_URI}`
+        docker cp /etc/${JUPYTER_NOTEBOOK_CONFIG} ${JUPYTER_SERVER_NAME}:${JUPYTER_HOME}/
+      fi
+
+      # If a Jupyter user script was specified, copy it into the jupyter docker container.
+      # TODO: not Jupyter specific
+      if [ ! -z ${JUPYTER_USER_SCRIPT_URI} ] ; then
+        log 'Installing Jupyter user extension [$JUPYTER_USER_SCRIPT_URI]...'
+        gsutil cp ${JUPYTER_USER_SCRIPT_URI} /etc
+        JUPYTER_USER_SCRIPT=`basename ${JUPYTER_USER_SCRIPT_URI}`
+        docker cp /etc/${JUPYTER_USER_SCRIPT} ${JUPYTER_SERVER_NAME}:${JUPYTER_HOME}/${JUPYTER_USER_SCRIPT}
+        retry 3 docker exec -u root ${JUPYTER_SERVER_NAME} chmod +x ${JUPYTER_HOME}/${JUPYTER_USER_SCRIPT}
+        retry 3 docker exec -u root -e PIP_USER=false ${JUPYTER_SERVER_NAME} ${JUPYTER_HOME}/${JUPYTER_USER_SCRIPT}
+      fi
+
+      log 'Starting Jupyter Notebook...'
+      retry 3 docker exec -d ${JUPYTER_SERVER_NAME} ${JUPYTER_SCRIPTS}/run-jupyter.sh
```

Comment:
I just remembered, we also install a startup script on the master node to restart Jupyter when the instance is restarted:
https://github.com/DataBiosphere/leonardo/blob/develop/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala#L117-L124

We should probably change this script to test if the Jupyter image is installed.

Note: RStudio doesn't need an explicit startup command because we just use the docker `ENTRYPOINT`.

Reply:
I don't think I saw this addressed -- the above script assumes the existence of Jupyter, so probably needs to change

Reply:
i did this in leo code instead of script https://github.com/DataBiosphere/leonardo/pull/718/commits/e72240b66372e1afc6fa0febbe1934fb557d364c#diff-9a3d2448dba603018e03e21dab912aa9R489

i think a follow-up to all of this would be to no longer hard code these things at all